### DISCUSSION TITLE
Case for NixOS

### DISCUSSION THREAD

**Discussion by ypid** [debops/debops#2528] [Ideas]
Why is NixOS awesome for me with an Ansible and DebOps background?

* [Most up-to-date operating system while also offering the most packages](https://www.linux-magazine.com/Online/News/NixOS-24.05-Is-Ready-for-Prime-Time). This makes it suitable as a one-stop shop for all my software/services. So no more Docker images to run random services. So I can standardize how I audit/build/run my software.
* When you create a file with Ansible and later remove that task, the file stays on the machine. Nix cleans up everything that is not defined in the config.
* It is way faster than Ansible (tested without changes). Running all Ansible roles against a typical server VM took ~10 minutes for me. Applying the NixOS config to a VM takes ~20 seconds for me.
* There are a lot more maintainers for roles I feel. For example, [Nextcloud is maintained without me doing anything](https://nixos.org/manual/nixos/stable/#module-services-nextcloud).
* The module system is ultra flexible. With Ansible, role users can only change parameters a role explicitly exposes using role variables. With Nix modules, users can hook in anywhere.
* With Ansible, it takes ~20 lines of template code plus ~4 tasks to define a systemd service and start it. With Nix modules, it only takes a few lines without any redundancy. With Ansible, you repeat the service name, in task name, task parameters and file names. Writing Ansible roles with all this redundancy and requirement to name tasks got less fun for me over the years. Why do I have to come up with task names again? Because Ansible is so slow that I can even read them :) Back to first principles. I want to configure a system, not come up with task names when what I am doing is self-explanatory.
* The whole OS is build by configuration management. Debian has packages that place default config files. Ontop of that, you have Ansible, which tries to hook into this default config somehow with all the troubles this has and is different for every service. With NixOS, you can just change the default config that NixOS provides. Less messing around.

Contra:

* No signed iso images available yet.
* Not everything is build from source. Some packages pull binaries or unpack deb packages.

See also https://me.ypid.de/timeline/. I am running Nextcloud with NixOS. Works very well.

I am currently using [Nix imports](https://nixos.org/manual/nixos/stable/#sec-modularity) as a replacement for the Ansible inventory. This is already more powerful than Ansible. I have one git repo that I sync to all the NixOS machines and then `nixos-rebuild switch` in that machine. I will checkout the tools for managing multiple NixOS machines later just to see how others are doing it.

Cc: @Erethon, @drybjed

**Comment by drybjed** [comment]
The most important thing is - how do we start?

Luckily, you can run NixOS installer right from the `iPXE` boot menu configured by DebOps. After loading the iPXE boot menu, go to "Netboot.xyz Boot menu" -> "Linux Network Installs (64-bit)" -> "NixOS" -> pick the preferred release, and the host should boot right into NixOS installer. Enjoy. :-)
